addons/udes_stock/models/stock_picking_batch.py

```diff
@@ -1017,21 +1017,43 @@ def reserve_pallet(self, pallet_name):
         Raises a ValidationError if the pallet is already reserved for another
         batch.
         """
+        Picking = self.env['stock.picking']
         PickingBatch = self.env['stock.picking.batch']
 
         self.ensure_one()
 
-        conflicting_batch = PickingBatch.search([
-            ('id', '!=', self.id),
-            ('state', '=', 'in_progress'),
-            ('u_last_reserved_pallet_name', '=', pallet_name),
-        ])
+        reserve_pallet_per_picking = self.picking_ids.filtered('picking_type_id.u_reserve_pallet_per_picking')
+
+        if reserve_pallet_per_picking:
+            conflicting_picking = Picking.search([
+                ('u_reserved_pallet', '=', pallet_name),
+                ('state', 'not in', ['draft', 'cancelled', 'done'])
+            ])
+            conflicting_batch = conflicting_picking.batch_id
+        else:
+            conflicting_batch = PickingBatch.search([
+                ('id', '!=', self.id),
+                ('state', '=', 'in_progress'),
+                ('u_last_reserved_pallet_name', '=', pallet_name),
+            ])
         if conflicting_batch:
             raise ValidationError(
                 _("This pallet is already being used for batch %s.") %
                 conflicting_batch[0].name)
 
-        self.write({ 'u_last_reserved_pallet_name': pallet_name })
+        if reserve_pallet_per_picking:
+            # We expect at least one picking has already been assigned to the
```

Comment:
Reserve pallet should either reserve a pallet for the batch (stored at `u_last_reserved_pallet_name`) or for the picking (stored at `u_reserved_pallet`). So at this point we should write to picking straightaway.
We can add a check at the start of the function to ensure we have a picking and it is in the batch in self.

Reply:
> We can add a check at the start of the function to ensure we have a picking

Are we assuming that the front end will always specify a picking if matrix picking is enabled? (@squah-u ?)

Reply:
The matrix pick flow will always specify a picking while the regular pick flow will never specify a picking. It would be technically possible for `u_reserve_pallet_per_picking` and the normal pick flow to be both enabled at the same time.